sw/device/silicon_creator/lib/dice.c

```diff
@@ -95,40 +98,15 @@ static void curr_pubkey_le_to_be_convert(attestation_public_key_t *pubkey) {
   le_be_buf_format((unsigned char *)pubkey->y, kAttestationPublicKeyCoordBytes);
 }
 
-rom_error_t dice_attestation_keygen(dice_key_t desired_key,
-                                    hmac_digest_t *pubkey_id,
-                                    attestation_public_key_t *pubkey) {
-  attestation_key_seed_t otbn_ecc_keygen_seed;
-  sc_keymgr_diversification_t keymgr_diversifier = {
-      .salt = {0},
-      .version = 0,
-  };
-
-  switch (desired_key) {
-    case kDiceKeyUds:
-      HARDENED_RETURN_IF_ERROR(
-          sc_keymgr_state_check(kScKeymgrStateCreatorRootKey));
-      otbn_ecc_keygen_seed = kUdsAttestationKeySeed;
-      keymgr_diversifier = kUdsKeymgrDiversifier;
-      break;
-    case kDiceKeyCdi0:
-      HARDENED_RETURN_IF_ERROR(
-          sc_keymgr_state_check(kScKeymgrStateOwnerIntermediateKey));
-      keymgr_diversifier = kCdi0KeymgrDiversifier;
-      otbn_ecc_keygen_seed = kCdi0AttestationKeySeed;
-      break;
-    case kDiceKeyCdi1:
-      HARDENED_RETURN_IF_ERROR(sc_keymgr_state_check(kScKeymgrStateOwnerKey));
-      keymgr_diversifier = kCdi1KeymgrDiversifier;
-      otbn_ecc_keygen_seed = kCdi1AttestationKeySeed;
-      break;
-    default:
-      return kErrorDiceInvalidArgument;
-  };
+static rom_error_t common_attestation_keygen(
+    const sc_keymgr_diversification_t *diversifier, attestation_key_seed_t seed,
+    sc_keymgr_state_t desired_keymgr_state, hmac_digest_t *pubkey_id,
+    attestation_public_key_t *pubkey) {
+  HARDENED_RETURN_IF_ERROR(sc_keymgr_state_check(desired_keymgr_state));
 
   // Generate / sideload key material into OTBN, and generate the ECC keypair.
-  HARDENED_RETURN_IF_ERROR(otbn_boot_attestation_keygen(
-      otbn_ecc_keygen_seed, keymgr_diversifier, pubkey));
+  HARDENED_RETURN_IF_ERROR(
+      otbn_boot_attestation_keygen(seed, *diversifier, pubkey));
```

Comment:
We will need to refactor the `otbn_boot_attestation_keygen` function to:
1. Allow selecting which side of the key ladder we are generating a key from. For the TPM key case we want to generate from the `sealing` side, but as is currently written, the `otbn_boot_attestation_keygen` function calls `sc_keymgr_generate_attestation_key_otbn` [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/otbn_boot_services.c;drc=30d7e787c753caaa03fe68a4a70da1bbcbc1d96f;l=132), which under the hood, selects the `attestation` side of the key ladder [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/keymgr.c;drc=30d7e787c753caaa03fe68a4a70da1bbcbc1d96f;l=216).
2. rename this function from `otbn_boot_attestation_keygen` to `otbn_boot_ecc_p256_keygen`, or something like that

I can turn this into a separate issue, and we can submit a follow up PR to address.

Reply:
Added an Issue to track this: https://github.com/lowRISC/opentitan/issues/22622